sequencer/src/block/full_payload/ns_proof.rs

```diff
@@ -0,0 +1,161 @@
+use crate::{
+    block::{
+        full_payload::{ns_table::NsTable, payload::Payload, payload::PayloadByteLen},
+        namespace_payload::NsPayloadOwned,
+    },
+    NamespaceId, Transaction,
+};
+use hotshot_types::{
+    traits::EncodeBytes,
+    vid::{vid_scheme, LargeRangeProofType, VidCommitment, VidCommon, VidSchemeType},
+};
+use jf_vid::{
+    payload_prover::{PayloadProver, Statement},
+    VidScheme,
+};
+use serde::{Deserialize, Serialize};
+
+/// Proof of correctness for namespace payload bytes in a block.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct NsProof {
+    ns_id: NamespaceId,
+    existence: Option<NsProofExistence>, // `None` if `ns_id` is not in the block.
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+struct NsProofExistence {
+    ns_payload: NsPayloadOwned,
+    ns_proof: LargeRangeProofType,
+}
+
+impl NsProof {
+    /// Returns the payload bytes for namespace `ns_id`, along with a proof of
+    /// correctness for those bytes. Returns `None` on error.
+    ///
+    /// The namespace payload is included as a hidden field in the returned
+    /// [`NsProof`]. A conventional API would instead return `(NsPayload,
+    /// NsProof)` and [`NsProof`] would not contain the namespace payload.
+    /// ([`TxProof::new`](super::tx_proof::TxProof::new) conforms to this
+    /// convention.) In the future we should change this API to conform to
+    /// convention. But that would require a change to our RPC endpoint API at
+    /// [`endpoints`](crate::api::endpoints), which is a hassle.
+    pub fn new(payload: &Payload, ns_id: NamespaceId, common: &VidCommon) -> Option<NsProof> {
+        let payload_byte_len = payload.byte_len();
+        if !payload_byte_len.is_consistent(common) {
+            return None; // error: vid_common inconsistent with self
+        }
+        let Some(ns_index) = payload.ns_table().find_ns_id(&ns_id) else {
+            // ns_id does not exist
+            return Some(NsProof {
```

Comment:
Should not we have some cryptographic evidence that the payload is empty for namespace `ns_id`?

Reply:
Here there is significant benefit if we take advantage of the fact that every node sees and comes to consensus on the namespace table: non-existence of a namespace ID can be verified directly by searching the namespace table.

Now that you mention it... it is kinda silly for `NsProof` to have a `existence` field, and pretend to "prove" non-existence of a namespace ID to a "verifier" who will just search the namespace table.

Perhaps I should tweak `NsProof` so that its constructor refuses to make a proof for a non-existent namespace ID. I think I'll do that next time I'm at my desk.

Reply:
Oh I see, good to know as it is relevant for the design of the zk rollups circuits.

Reply:
> Perhaps I should tweak `NsProof` so that its constructor refuses to make a proof for a non-existent namespace ID. I think I'll do that next time I'm at my desk.

done in https://github.com/EspressoSystems/espresso-sequencer/pull/1499/commits/f907be50155e5ff42fafc5840e4bd96752ebedd8